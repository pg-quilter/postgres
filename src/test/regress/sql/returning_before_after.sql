--
-- Test BEFORE/AFTER feature in RETURNING statements

CREATE TABLE foo (
		bar1 INTEGER,
		bar2 TEXT
		);

INSERT INTO foo VALUES (1, 'x'),(2,'y');

UPDATE foo SET bar1=bar1+1 RETURNING before.*, bar1, bar2;

UPDATE foo SET bar1=bar1-1 RETURNING after.bar1, before.bar1*2;

UPDATE foo SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*;


-- it should fail
UPDATE foo SET bar1=bar1+before.bar1 RETURNING before.*;
UPDATE foo SET bar1=bar1+after.bar1 RETURNING after.*;

-- test before/after aliases
UPDATE foo AS before SET bar1=bar1+1 RETURNING before.*,after.*;
UPDATE foo AS after SET bar1=bar1-1 RETURNING before.*,after.*;

-- test inheritance
CREATE TABLE foo2 (bar INTEGER) INHERITS(foo);

INSERT INTO foo2 VALUES (1,'b',5);

UPDATE foo2 SET bar1=bar1*2, bar=bar1+5, bar2=bar1::text || bar::text RETURNING before.*, after.*, *;

-- check views

CREATE VIEW view_foo AS SELECT * FROM foo;

UPDATE foo SET bar1=bar1+1 RETURNING before.*, bar1, bar2;

CREATE TABLE foo3 (bar1 INTEGER, bar4 FLOAT);

INSERT INTO foo2 VALUES (2, 'asdf', 33);
INSERT INTO foo3 VALUES (2, 7.77);

CREATE VIEW view_join AS SELECT f2.*, f3.bar1 AS f1bar1, f3.bar4 FROM foo2 f2 
JOIN foo3 f3 ON f2.bar1 = f3.bar1;

UPDATE view_join SET bar1=bar1+5, bar2=bar2||'join', bar=bar1*2, bar4=7 RETURNING before.*, after.*;

-- check triggers
CREATE FUNCTION returning_trig() returns trigger as $$
BEGIN
NEW.bar1 = NEW.bar1*NEW.bar1; 
RETURN NEW;
END; $$ language plpgsql;

DROP TABLE foo2 CASCADE;
CREATE TRIGGER bef_foo BEFORE UPDATE ON foo FOR EACH ROW EXECUTE PROCEDURE returning_trig();

UPDATE foo SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*, *;

DROP TABLE foo CASCADE;
DROP TABLE foo3 CASCADE;


