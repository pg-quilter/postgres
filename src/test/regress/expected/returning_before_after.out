--
-- Test BEFORE/AFTER feature in RETURNING statements
CREATE TABLE foo (
		bar1 INTEGER,
		bar2 TEXT
		);
INSERT INTO foo VALUES (1, 'x'),(2,'y');
UPDATE foo SET bar1=bar1+1 RETURNING before.*, bar1, bar2;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    1 | x    |    2 | x
    2 | y    |    3 | y
(2 rows)

UPDATE foo SET bar1=bar1-1 RETURNING after.bar1, before.bar1*2;
 bar1 | ?column? 
------+----------
    1 |        4
    2 |        6
(2 rows)

UPDATE foo SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    1 | x    |    2 | xz
    2 | y    |    3 | yz
(2 rows)

-- it should fail
UPDATE foo SET bar1=bar1+before.bar1 RETURNING before.*;
ERROR:  missing FROM-clause entry for table "before"
LINE 1: UPDATE foo SET bar1=bar1+before.bar1 RETURNING before.*;
                                 ^
UPDATE foo SET bar1=bar1+after.bar1 RETURNING after.*;
ERROR:  missing FROM-clause entry for table "after"
LINE 1: UPDATE foo SET bar1=bar1+after.bar1 RETURNING after.*;
                                 ^
-- test before/after aliases
UPDATE foo AS before SET bar1=bar1+1 RETURNING before.*,after.*;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    3 | xz   |    3 | xz
    4 | yz   |    4 | yz
(2 rows)

UPDATE foo AS after SET bar1=bar1-1 RETURNING before.*,after.*;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    3 | xz   |    2 | xz
    4 | yz   |    3 | yz
(2 rows)

-- test inheritance
CREATE TABLE foo2 (bar INTEGER) INHERITS(foo);
INSERT INTO foo2 VALUES (1,'b',5);
UPDATE foo2 SET bar1=bar1*2, bar=bar1+5, bar2=bar1::text || bar::text RETURNING before.*, after.*, *;
 bar1 | bar2 | bar | bar1 | bar2 | bar | bar1 | bar2 | bar 
------+------+-----+------+------+-----+------+------+-----
    1 | b    |   5 |    2 | 15   |   6 |    2 | 15   |   6
(1 row)

-- check views
CREATE VIEW view_foo AS SELECT * FROM foo;
UPDATE foo SET bar1=bar1+1 RETURNING before.*, bar1, bar2;
 bar1 | bar2 | bar1 | bar2 
------+------+------+------
    2 | xz   |    3 | xz
    3 | yz   |    4 | yz
    2 | 15   |    3 | 15
(3 rows)

CREATE TABLE foo3 (bar1 INTEGER, bar4 FLOAT);
INSERT INTO foo2 VALUES (2, 'asdf', 33);
INSERT INTO foo3 VALUES (2, 7.77);
CREATE VIEW view_join AS SELECT f2.*, f3.bar1 AS f1bar1, f3.bar4 FROM foo2 f2 
JOIN foo3 f3 ON f2.bar1 = f3.bar1;
UPDATE view_join SET bar1=bar1+5, bar2=bar2||'join', bar=bar1*2, bar4=7 RETURNING before.*, after.*;
ERROR:  cannot update view "view_join"
DETAIL:  Views that do not select from a single table or view are not automatically updatable.
HINT:  To make the view updatable, provide an unconditional ON UPDATE DO INSTEAD rule or an INSTEAD OF UPDATE trigger.
-- check triggers
CREATE FUNCTION returning_trig() returns trigger as $$
BEGIN
NEW.bar1 = NEW.bar1*NEW.bar1; 
RETURN NEW;
END; $$ language plpgsql;
DROP TABLE foo2 CASCADE;
NOTICE:  drop cascades to view view_join
CREATE TRIGGER bef_foo BEFORE UPDATE ON foo FOR EACH ROW EXECUTE PROCEDURE returning_trig();
UPDATE foo SET bar1=bar1+1, bar2=bar2 || 'z' RETURNING before.*, after.*, *;
 bar1 | bar2 | bar1 | bar2 | bar1 | bar2 
------+------+------+------+------+------
    3 | xz   |    4 | xzz  |   16 | xzz
    4 | yz   |    5 | yzz  |   25 | yzz
(2 rows)

DROP TABLE foo CASCADE;
NOTICE:  drop cascades to view view_foo
DROP TABLE foo3 CASCADE;
